### PR TITLE
Deal gracefully with symlinks

### LINKED ISSUE #57: Symlink in ignored directory causing out of memory error
## Bug Report

- [x] Yes, I reviewed the [contribution guidelines](https://make.wordpress.org/cli/handbook/contributing/).
- [x] Yes, more specifically, I reviewed the guidelines on [how to write clear bug reports](https://make.wordpress.org/cli/handbook/bug-reports/).

**Describe the current, buggy behavior**

I have a project structure with a recursive symlink:

```
project-dir/my-plugin.php
project-dir/wp-content/plugins/my-plugin/
```

where `wp-content/plugins/my-plugin/` is a symlink to `project-dir/`.

Previously I was using `dist-archive` on the `src` directory and now I'm trying to use it on the entire project directory, with appropriate an `.distignore` in the project root. [PhpStorm project screenshot](https://brianhenryie.s3.amazonaws.com/2021/wp-plugin-project-structure.png)

My `.distignore` file contains `wp-content` in it.

Still, running `wp dist-archive . out.zip` causes a `PHP Fatal error`:

```
PHP Fatal error:  Allowed memory size of 134217728 bytes exhausted (tried to allocate 121634848 bytes) in /path/to/project-dir/vendor/wp-cli/wp-cli/php/WP_CLI/Process.php on line 75
```

When I delete the symlink, I am able to use `dist-archive` without issue.

**Describe what you would expect as the correct outcome**

I expect the `wp-content` folder to have no bearing on the operation because it is listed in the `.distignore` file.

**Let us know what environment you are running this on**

MacOS M1,  Monterey (12.3.1), PHP 8.0 via Homebrew, Zsh shell.

```
OS:     Darwin 21.4.0 Darwin Kernel Version 21.4.0: Fri Mar 18 00:47:26 PDT 2022; root:xnu-8020.101.4~15/RELEASE_ARM64_T8101 arm64
Shell:  /bin/zsh
PHP binary:     /opt/homebrew/bin/php
PHP version:    8.0.19
php.ini used:   /opt/homebrew/etc/php/8.0/php.ini
MySQL binary:   /opt/homebrew/bin/mysql
MySQL version:  mysql  Ver 15.1 Distrib 10.7.3-MariaDB, for osx10.17 (arm64) using  EditLine wrapper
SQL modes:      STRICT_TRANS_TABLES,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION
WP-CLI root dir:        /Users/brianhenry/Sites/bh-wc-disable-free-gifts-per-product/vendor/wp-cli/wp-cli
WP-CLI vendor dir:      /Users/brianhenry/Sites/bh-wc-disable-free-gifts-per-product/vendor
WP_CLI phar path:       
WP-CLI packages dir:    /Users/brianhenry/.wp-cli/packages/
WP-CLI global config:   
WP-CLI project config:  /Users/brianhenry/Sites/bh-wc-disable-free-gifts-per-product/wp-cli.yml
WP-CLI version: 2.6.0

```

**Provide a possible solution**

Using the `zip` flag:

```
-y   store symbolic links as the link instead of the referenced file
```

Change:

`$cmd = "zip -r '{$archive_file}' {$archive_base} {$excludes}";`

to:

`$cmd = "zip -y -r '{$archive_file}' {$archive_base} {$excludes}";`

This works.

But I presume it would prevent the `dist-ignore` command from working with plugins that use symlinks internally to the plugin (i.e. not just dev tools and environment) during development.

Alternatively, I could copy all files to system temp before archiving them.

### PR DESCRIPTION
**Pull request by BrianHenryIE**
~~Needs tests.~~

Has a nice test that fails on main and passes with this PR.

Fixes #57

### CODE REVIEW

File: features/dist-archive.feature [hunk 1]
```diff
@@ -224,3 +224,37 @@ Feature: Generate a distribution archive of a project
     And the wp-content/plugins/hello-world/hello-world.php file should exist
     And the wp-content/plugins/hello-world/.travis.yml file should not exist
     And the wp-content/plugins/hello-world/bin directory should not exist
+
+Scenario: Avoids recursive symlink
+    Given a WP install in wordpress
+    And a .distignore file:
+      """
+      wp-content
+      wordpress
```
Comment (schlessera) [inline features/dist-archive.feature:233]: The tests should trigger the extra regex rules as well to ensure they work as expected:
- using an asterisk
- using a slash
- using a dot

**Review by schlessera — CHANGES_REQUESTED** [review]
Code looks,

File: src/Dist_Archive_Command.php [hunk 2]
```diff
@@ -119,7 +120,77 @@ public function __invoke( $args, $assoc_args ) {
...
+
+			$iterator = new RecursiveIteratorIterator(
+				new RecursiveDirectoryIterator( $path, RecursiveDirectoryIterator::SKIP_DOTS ),
+				RecursiveIteratorIterator::SELF_FIRST
+			);
+
+			/**
+			 * @var RecursiveIteratorIterator $iterator
+			 * @var SplFileInfo $item
+			 */
```
[the remark this answers is not in the document]
Comment (BrianHenryIE) [reply src/Dist_Archive_Command.php:147]: This was needed in PhpStorm. It was not auto-completing the `$item` class without it.

**Comment by schlessera** [comment]
@BrianHenryIE Now that I'm looking at the code again with fresh eyes... is there a reason why you're using closures in the first place? These two closures should rather be private methods of the class...

**Comment by BrianHenryIE** [comment]
Not particularly. Probably because there's no PHPUnit tests leading me to need them be callable functions. 

Why `private` instead of `protected`? `static` still?

**Comment by danielbachhuber** [comment]
Ship #61

**Pull request closed without merging**